### DISCUSSION TITLE
Section 5 lesson 128 getting pred_classes tensor of all 2s after using argmax(dim=1)

### DISCUSSION THREAD

**Discussion by AxisMeetsWorld** [mrdbourke/pytorch-deep-learning#500] [Q&A]
Hi @mrdbourke,
      I was trying to follow along with the above mentioned exercise and I tried to look at what you had [here](https://www.learnpytorch.io/03_pytorch_computer_vision/) and I couldn't figure out why after calling the following lines I was getting nothing but a tensor of 2s as opposed to what you got:
```
# Convert prediction probabilities to labels

pred_classes = pred_probs.argmax(dim=1)

pred_classes
```

The result was: **tensor([2, 2, 2, 2, 2, 2, 2, 2, 2])**
where you seem to get **tensor([5, 1, 7, 4, 3, 0, 4, 7, 1])** 😬 

In the previous cells, I ran the following code, and I don't think there is anything different:
``` def make_predictions(model: torch.nn.Module,
                     data: list,
                     device: torch.device  = device):
  pred_probs = []
  model.eval()
  with torch.inference_mode():
    for sample in data:
      #Prepare the sample
      sample = torch.unsqueeze(sample,dim =0).to(device)

      #Forward pass
      pred_logit = model(sample)

      # Get the prediction probability (logit -> prediction)
      pred_prob = torch.softmax(pred_logit.squeeze(), dim = 0)

      #Get the pred_prob off the GPU for further calculations
      pred_probs.append(pred_prob.cpu())

  #Stack the pred_probs  to turn the list into a tensor
  return torch.stack(pred_probs)

```
The next cell

```
import random
random.seed(42)
test_samples =[]
test_labels = []
for sample, label in random.sample(list(test_data), k=9):
  test_samples.append(sample)
  test_labels.append(label)
print(f"Test sample image shape: {test_samples[0].shape}\nTest sample label: {test_labels[0]} ({class_names[test_labels[0]]})")
```

Finally right before I called the suspect code (giving the 2s) I had

```
# Make some predictions
pred_probs = make_predictions(model= model_2, 
                              data=test_samples)

pred_probs[:2] 

```
Any ideas on what I might be missing so that I can get some semblance of what you had? :
**tensor([5, 1, 7, 4, 3, 0, 4, 7, 1])** as opposed to **tensor([2, 2, 2, 2, 2, 2, 2, 2, 2])**

**Comment by AxisMeetsWorld** [comment]
Hi again,
      I tried to look more at this to see if I could possibly figure out what I did differently, and I can't seem to figure it out. I tried to run a squeeze on both the `test_samples` and the `pred_probs` and that resulted in errors. I went back and tried to watch what was done in the video and when I tested what `pred_probs[:10]` looked  like here is what I saw:
**tensor([[0.1030, 0.0933, 0.1060, 0.0963, 0.0985, 0.1013, 0.0995, 0.0991, 0.1020,
         0.1010],
        [0.1040, 0.0933, 0.1045, 0.0967, 0.0991, 0.1018, 0.0994, 0.0988, 0.1025,
         0.0999],
        [0.1030, 0.0935, 0.1057, 0.0968, 0.0980, 0.1013, 0.0994, 0.0993, 0.1019,
         0.1010],
        [0.1034, 0.0935, 0.1054, 0.0962, 0.0992, 0.1016, 0.0994, 0.0987, 0.1022,
         0.1003],
        [0.1036, 0.0931, 0.1050, 0.0961, 0.0994, 0.1015, 0.1000, 0.0981, 0.1028,
         0.1004],
        [0.1036, 0.0937, 0.1048, 0.0964, 0.0989, 0.1017, 0.0993, 0.0982, 0.1028,
         0.1006],
        [0.1034, 0.0936, 0.1054, 0.0962, 0.0988, 0.1012, 0.0997, 0.0989, 0.1027,
         0.1002],
        [0.1032, 0.0938, 0.1050, 0.0974, 0.0983, 0.1014, 0.0991, 0.0986, 0.1023,
         0.1008],
        [0.1033, 0.0934, 0.1044, 0.0965, 0.0997, 0.1016, 0.0994, 0.0985, 0.1029,
         0.1002]])**
So it looks like somehow these numbers aren't quite matching what was in the video, but what it did show was the dimensionality being the same, and oddly the randomness did seem to show the index of 2 always being the greatest. Is there anything with my settings in the ranodom_seed() that could be off at all? At this point I am sort of at a loss for what might be causing this error.
I am also noticing those "2" values in the next section result in constantly getting a "pullover" prediction for the title plots (regardless of plot), so if I can figure out what I am doing wrong that is getting all of those "2" values, it will help with the rest of this chapter. Thanks!

**Comment by mrdbourke** [comment]
Hi @AxisMeetsWorld , 

These errors are tricky to track down.

I'd say it could be your random seed setup or how your model is learning (in this case, I'd check the forward/backward pass/loss values during training/testing).

I just ran the [03 notebook](https://www.learnpytorch.io/03_pytorch_computer_vision/) end-to-end and got the same original results: https://colab.research.google.com/drive/17uginhratDnDw9c10iH51molxbJprg9m?usp=sharing 

```python
# Turn the prediction probabilities into prediction labels by taking the argmax()
pred_classes = pred_probs.argmax(dim=1)
pred_classes
```

Output: 

```
tensor([5, 1, 7, 4, 3, 0, 4, 7, 1])
```

It will also sometimes be dependent on what hardware you're using (though not *too* dependent) - are you running it in Google Colab?

**Reply by AxisMeetsWorld** [reply]
Thanks for the response @mrdbourke,
       I am trying to recall, is there some setting that I am forgetting for a random seed other than the `random.seed(42)`? 
If my understanding of the `argmax(dim=1)` call is correct, it is taking the index of the largest value in the **inside** of the **pred_probs** tensor. In the sample above that seems consistent, so I definitely see where the randomness might be a little off here.
Wondering if there is any setting like what I read about [here](https://vandurajan91.medium.com/random-seeds-and-reproducible-results-in-pytorch-211620301eba) where you can set some sort of` torch.get_rng_state()` to set a range? It seems that `test_samples` might somehow not be generating the right way in that for loop. Outside of that, I the rest of the Section 5 code worked as expected, so it just seems odd this one thing isn't.

**Reply by AxisMeetsWorld** [reply]
I should probably ask if there are any suggested troubleshooting steps I can run that I haven't which might help track this down?
